crates/revm/src/state_change.rs

```diff
@@ -46,6 +51,68 @@ pub fn post_block_balance_increments(
     balance_increments
 }
 
+/// Applies the pre-block call to the EIP-4788 beacon block root contract, using the given block,
+/// [ChainSpec], EVM.
+///
+/// If cancun is not activated or the block is the genesis block, then this is a no-op, and no
+/// state changes are made.
+#[inline]
+pub fn apply_beacon_root_contract_call<DB: Database>(
+    chain_spec: &ChainSpec,
+    block_timestamp: u64,
+    block_number: u64,
+    block_parent_beacon_block_root: Option<H256>,
+    evm: &mut EVM<State<DB>>,
+) -> Result<(), BlockExecutionError>
+where
+    <DB as Database>::Error: Debug,
+{
+    if chain_spec.is_cancun_activated_at_timestamp(block_timestamp) {
+        // if the block number is zero (genesis block) then the parent beacon block root must
+        // be 0x0 and no system transaction may occur as per EIP-4788
+        if block_number == 0 {
+            if block_parent_beacon_block_root != Some(H256::zero()) {
+                return Err(BlockValidationError::CancunGenesisParentBeaconBlockRootNotZero.into())
+            }
+        } else {
+            let parent_beacon_block_root = block_parent_beacon_block_root.ok_or(
+                BlockExecutionError::from(BlockValidationError::MissingParentBeaconBlockRoot),
+            )?;
+
+            // get previous env
+            let previous_env = evm.env.clone();
```

Comment:
nit: can you not instead call this `beacon_roots_env` and do the `fill_tx` call with that, to avoid having to re-set the `evm.env` at the end of the function?

Reply:
no, because the env would have to be set in the evm anyways, for it to be used during `transact`. We could initialize a new `Env` and `swap` them, but I don't see how that's different than cloning here, since we have to initialize a second `Env` anyways.

We could get rid of the reset by just telling the caller (in docs) to make sure the block and tx env are initialized properly before the next call to `transact`, and make sure to do that in the processor, payload builder, etc.